galsim/detectors.py

```diff
@@ -275,6 +275,62 @@ def applyIPC(self, IPC_kernel, edge_treatment='extend', fill_value=None, kernel_
     else:
         self.array[:,:] = out_array
 
+def applyPersistence(self,imgs,coeffs):
+    """
+    Applies the effects of persistence to the Image instance.
+
+    Persistence refers to the retention of a small fraction of the signal after resetting the
+    imager pixel elements. The persistence signal of a previous exposure is left in the pixel even
+    after several detector resets. This effect is most likely due to charge traps in the material.
+    Laboratory tests on the WFIRST CMOS detectors show that if exposures and readouts are taken in
+    a fixed cadence, the persistence signal can be given as a linear combination of prior pixel
+    values that can be added to the current image.
+
+    This routine takes in one or multiple Image instances and adds them to Image weighted by the
+    values passed on to 'coeffs'. This routine does NOT keep track of realistic dither patterns.
+    During the image simulation process, the user has to queue a list of previous Image
+    instances (img_list) outside the routine by inserting the latest image in the beginning of the
+    list and deleting the oldest image. The values in 'coeffs' tell how much of each Image is to
+    be added. This usually remains constant in the image generation process.
+
+    Calling
+    -------
+
+        >>> img.applyPersistence(imgs=img_list, coeffs=coeffs_list)
+
+    @ param imgs:       an Image instance or a list of previous Image instances that still persist
+    @ param coeffs:     specifies the retention factor for one or all of the Image instances listed
+                        in 'imgs'. Different retention factors for different instances can be
+                        specified as a list or tuples in the same order as that of the images.
+
+    @ returns None
+    """
+
+    if isinstance(imgs,galsim.Image) and (isinstance(coeffs,float) or isinstance(coeffs,int)):
```

Comment:
Well, the only (lame) reason I had it there because if one of the coeffs is
simply 0, it would not be considered as a float. There is no reason I can
think of as to why 0 would be passed, so I could remove that condition.

On Wed, Apr 20, 2016 at 7:42 AM, Rachel Mandelbaum <notifications@github.com
> wrote:

> In galsim/detectors.py
> <https://github.com/GalSim-developers/GalSim/pull/710#discussion_r60392494>
> :
>
> > +    be added. This usually remains constant in the image generation process.
> > +
> > +    Calling
> > +    -------
> > +
> > +        >>> img.applyPersistence(imgs=img_list, coeffs=coeffs_list)
> > +
> > +    @ param imgs:       an Image instance or a list of previous Image instances that still persist
> > +    @ param coeffs:     specifies the retention factor for one or all of the Image instances listed
> > +                        in 'imgs'. Different retention factors for different instances can be
> > +                        specified as a list or tuples in the same order as that of the images.
> > +
> > +    @ returns None
> > +    """
> > +
> > +    if isinstance(imgs,galsim.Image) and (isinstance(coeffs,float) or isinstance(coeffs,int)):
>
> Do you really think that coeffs could be an int and it wouldn't just be a
> bug?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/GalSim-developers/GalSim/pull/710/files/cc5ed0d33b54e8526cc73690e18cc672c180a301#r60392494>
>



-- 
Regards

Arun Kannawadi Jayaraman
PhD candidate, Dept. of Physics,
Carnegie Mellon University,
Pittsburgh, PA 15213